README.md
# How to tweet your releases
> Send a tweet to let people know when you release software on GitHub.


It's great to release software. And even better when people actually know about it!

One way to let people know about your new software releases is to tell them on Twitter. But we're software developers, not social media managers -- so that means we *automate all the things*.

With tweetrel, you can have GitHub send a tweet for you whenever you make a release.

## Install

`pip install tweetrel`

## How to use

First, you need to add your Twitter API details as [GitHub secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets). You'll need to sign up for a [Twitter API](https://developer.twitter.com/en/docs) account if you don't already have one. Once you've signed up, Twitter will give you four keys: `consumer_key`, `consumer_secret`, `access_token`, and `access_token_secret`. Create a new secret called `TWITTER`, and paste those four keys in, with a space between each one.

If you need to send Tweets on behalf of a different user to the one attached to your login, you'll need to authenticate with Twitter as that user. Run `tweetrel-auth` in your terminal and follow the instructions, and it will give you the `access_token` and `access_token_secret` you need -- paste them (after a space) after your `consumer_key` and `consumer_secret` and that will let you send tweets as the user you authenticated as.

Next, in your terminal, `cd` to the root of your repo, and then run:

```bash
tweetrel-install
```

That will set `tweetrel` up for you to run automatically upon release. You'll see two folders added to your `.github` folder, containing the YAML workflow and python script. No additional setup is needed, other than pushing these to GitHub:

```bash
git add -A .github
git commit -am 'Add tweetrel`
git push
```

You can test this without actually making a release by going to your GitHub repo in your browser, clicking the *Actions* tab, then clicking on the *tweet* action, and clicking the *run workflow* button. That will use GitHub's example release payload for testing.

Once you've confirmed it's working, try making a release on GitHub, and check that you see the green tick in the Actions tab showing a successful run of the *tweet* action.

## Customization

As you'll see, the default template is:

    New #{repo} release: v{tag_name}. {html_url}\n\n{body}

You can customize that by setting the `TWEETREL_TEMPLATE` environment variable. For instance, add this (with suitable indentation) to the `env:` section of your workflow YAML file:

```bash
TWEETREL_TEMPLATE: |
  We are proud to announce the {tag_name} release of {repo}.
  
  For more, see: {html_url}. Here are the details:
  {body}
```

The `|` is the YAML character that indicates a multiline string. The sections in `{...}` are the Python formatting template variables that will be replaced by the details of your release. You don't need to include all of them if you don't want them.


CHANGELOG.md
# Release notes

<!-- do not remove -->

## 0.0.5

- Initial release



.github/scripts/build-tweet-release.py
from fastcore.all import *
from ghapi import *
import tweepy

def twitter_api():
    consumer_key,consumer_secret,access_token,access_token_secret = context_secrets.TWITTER.split()
    auth = tweepy.OAuthHandler(consumer_key,consumer_secret)
    auth.set_access_token(access_token,access_token_secret)
    return tweepy.API(auth)

def tweet_text(payload):
    rel = payload.release
    owner,repo = re.findall(r'https://api.github.com/repos/([^/]+)/([^/]+)/', rel.url)[0]
    tweet_tmpl = "New #{repo} release: v{tag_name}. {html_url}\n\n{body}"
    res = tweet_tmpl.format(repo=repo, tag_name=rel.tag_name, html_url=rel.html_url, body=rel.body)
    if len(res)<=280: return res
    return res[:279] + "…"

def send_tweet():
    payload = context_github.event
    if 'workflow' in payload: payload = example_payload(Event.release)
    if payload.action == 'published': return twitter_api().update_status(tweet_text(payload))

send_tweet()



tweetrel/core.py
# AUTOGENERATED! DO NOT EDIT! File to edit: 00_core.ipynb (unless otherwise specified).

__all__ = ['auth', 'twitter_api', 'example_rel', 'tweet_text', 'send_tweet', 'install']

# Cell
from fastcore.utils import *
from fastcore.script import *

# Cell
import tweepy

# Cell
@call_parse
def auth():
    "Function for `tweetrel-auth` CLI command"
    cons = input('Enter consumer key and secret, separated by a space: ').split()
    auth = tweepy.OAuthHandler(*cons)
    print('Go to this URL and login:')
    print(auth.get_authorization_url())
    code = input('Paste code here: ')
    print('Your keys are:')
    print(' '.join(auth.get_access_token(code)))

# Cell
from ghapi import *

# Cell
def twitter_api():
    consumer_key,consumer_secret,access_token,access_token_secret = context_secrets.TWITTER.split()
    auth = tweepy.OAuthHandler(consumer_key,consumer_secret)
    auth.set_access_token(access_token,access_token_secret)
    return tweepy.API(auth)

# Cell
example_rel = example_payload(Event.release)

# Cell
def tweet_text(payload):
    if 'workflow' in payload: payload = example_rel
    def_tmpl = "New #{repo} release: v{tag_name}. {html_url}\n\n{body}"
    tweet_tmpl = os.getenv('TWEETREL_TEMPLATE', def_tmpl)
    rel = payload.release
    owner,repo = payload.repository.full_name.split('/')
    res = tweet_tmpl.format(repo=repo, tag_name=rel.tag_name, html_url=rel.html_url, body=rel.body)
    if len(res)<=280: return res
    return res[:279] + "…"

# Cell
def send_tweet():
    payload = context_github.event
    if 'workflow' in payload: payload = example_payload(Event.release)
    if payload.action == 'published': return twitter_api().update_status(tweet_text(payload))

# Cell
@call_parse
def install():
    fill_workflow_templates(
        name='tweet', event="release:\n  types: [published]",
        run='pip install -Uq tweetrel',
        context=env_contexts('secrets'),
        script="import tweetrel\ntweetrel.send_tweet()"
    )